### PR TITLE
lib: location: Improve LTE-GNSS interworking and GNSS priority mode

### PR DESCRIPTION
**Pull request by hiltunent**
Added feature to location lib which triggers GNSS priority mode if 5 consecutive PVT events indicate that GNSS is blocked by LTE. This feature is also taken into use in Asset Tracker v2. Moreover, fixed a minor bug that prohibited GNSS from being started in case LTE is enabled, but coverage at startup is so bad that RRC connection establishment is not even attempted.

Signed-off-by: Tuomas Hiltunen <tuomas.hiltunen@nordicsemi.no>

### CODE REVIEW

File: lib/location/method_gnss.c
Comment (hiltunent) [inline lib/location/method_gnss.c:90]: This is the part that fixes GNSS not started issue in very bad LTE coverage. All the other changes are related to priority mode.

File: lib/location/method_gnss.c
Comment (trantanen) [inline lib/location/method_gnss.c:113]: I'd like to see a comment for this parameter as it's not evident.
I'm also thinking about the name as it doesn't ring a bell on what it would be. I'm thinking now `gnss_blocked_count` etc. but please discuss with @tokangas about the proper name. Maybe the current one is something you've already agreed on.

File: lib/location/method_gnss.c
Comment (trantanen) [inline lib/location/method_gnss.c:696]: Maybe a comment on top of the if-statement on what is done here just like there is in the previous if-statement.

File: applications/asset_tracker_v2/src/modules/location_module.c
Comment (trantanen) [inline applications/asset_tracker_v2/src/modules/location_module.c:272]: Move this 10 lines above. PR #8631 added that branch.

Comment (trantanen) [inline applications/asset_tracker_v2/src/modules/location_module.c:272]: And I want to confirm from @simensrostad that this is desired in ATv2. Read the description in location.h configuration description.

File: include/modem/location.h
Comment (tokangas) [inline include/modem/location.h:272]: There are two different situations which should be clearly separated:

- GNSS is totally blocked and does not get runtime at all (RRC connected mode).
- GNSS may be getting almost all runtime, but it's getting constantly interrupted by LTE (RRC idle mode).

It should be clear that this feature is for the latter case, thus I would rephrase the "does not get enough runtime" part.

File: include/modem/location.h
Comment (tokangas) [inline include/modem/location.h:275]: Same comment here for "blocked by LTE operation".

File: lib/location/method_gnss.c
Comment (tokangas) [reply lib/location/method_gnss.c:113]: The word "epoch" is not used elsewhere in our libraries or documentation, so it's probably not evident for everybody what it means (although it's a valid GNSS term). And again I'm slightly reserved about the word "blocked". 😄

File: lib/location/method_gnss.c
Comment (tokangas) [inline lib/location/method_gnss.c:614]:
```suggestion
	LOG_DBG("Tracked satellites: %d, fix valid: %s, insuf. time window: %s",
```

File: include/modem/location.h
Comment (jhirsi) [reply include/modem/location.h:272]: I was wondering what are the risks when enabling this? Is there a risk of missing important LTE data? i.e. mqtt cloud connection events or even the a-gps data? If yes, should this feature be used only when not (yet) connected (not registered and no PDN active) to LTE? 
Or, enable it only when no a-gps/p-gps phase currently on going?

File: lib/location/method_gnss.c
Comment (hiltunent) [reply lib/location/method_gnss.c:113]: Renamed to insuf_timewin_count and added comment.

File: include/modem/location.h
Comment (hiltunent) [reply include/modem/location.h:272]: It seems that the priority time window is short enough (40 seconds) that no data is lost. I tested this by triggering MO data (ping) from from Mosh right after requesting the priority window. The ping got sent successfully upon termination of the priority window. I tested MT data similarly with Contabo by scheduling a downlink data transmission to take place right after starting the priority mode. The data was received successfully when priority mode ended.

File: applications/asset_tracker_v2/src/modules/location_module.c
Comment (jhirsi) [inline applications/asset_tracker_v2/src/modules/location_module.c:272]: Not very familiar with ATv2 logics, but is this settable from the cloud plain like this?
Or would this priority_mode be enough to be Kconfig in location lib?

Comment (trantanen) [reply applications/asset_tracker_v2/src/modules/location_module.c]: We cannot override this at all currently from ATv2 cloud configuration, ATv2 Kconfig or loclib Kconfig.
Related to this comment:
https://github.com/nrfconnect/sdk-nrf/pull/9325#discussion_r1025595333
We had a discussion with @simensrostad on Friday and currently I don't expect this feature to be turned on for ATv2 in NCS 2.2.0. That's my guess based on what I've heard.

Generally about the configurability of certain loclib features, I think we should work on them after NCS 2.2.0 on a broader scope than just one specific item.

Comment (trantanen) [reply applications/asset_tracker_v2/src/modules/location_module.c]: @simensrostad Should we remove this line from ATv2 in NCS 2.2.0 context?

Comment (simensrostad) [reply applications/asset_tracker_v2/src/modules/location_module.c]: Hi, yes I think we should. We need to test and figure out the possible shortcomings of enabling GNSS prio in the application before we can comfortably enable it.

Comment (hiltunent) [reply applications/asset_tracker_v2/src/modules/location_module.c]: Fine, I'll remove it.

File: lib/location/method_gnss.c
Comment (tokangas) [inline lib/location/method_gnss.c:722]:
```suggestion
		/* GNSS must indicate that it is not getting long enough time windows for 5 consecutive
		 * epochs to trigger priority mode.
		 */
```

File: doc/nrf/releases/release-notes-changelog.rst
Comment (tokangas) [inline doc/nrf/releases/release-notes-changelog.rst:510]:
```suggestion
    * Improved LTE - GNSS interworking and added possibility to trigger GNSS priority mode if GNSS does not get long enough time windows due to LTE idle mode operations.
```

File: applications/asset_tracker_v2/src/modules/location_module.c
Comment (jhirsi) [reply applications/asset_tracker_v2/src/modules/location_module.c]: About loclib features and their configurability/usage from api, I see it as feature specific what are needed on api, what not, for example ncellmeas search types (new GCI search typed) are making very much impact on runtime, based on situation. Based on this to be removed code here, this particular feature looked like that runtime configuration isn't needed and thus was wondering why this was put to api and not as a Kconfig item?

File: include/modem/location.h
Comment (trantanen) [inline include/modem/location.h:280]: We could add something about the data transfer expectations. Something like:

If data is sent during the priority mode, it will be buffered and sent after the priority mode has finished.
If data would be available for receiving during the priority mode, it could be buffered by the network and sent to the device after the priority mode has finished. However, it is possible that network drops the data, or some protocol timer expires causing data transfer to stop.

@tokangas @jhirsi

File: applications/asset_tracker_v2/src/modules/location_module.c
Comment (trantanen) [reply applications/asset_tracker_v2/src/modules/location_module.c]: > this particular feature looked like that runtime configuration isn't needed and thus was wondering why this was put to api and not as a Kconfig item?

I guess to follow the general design of having configurability in the API's config structures. Whether this is a good design as such is another thing but that's how other features exist in GNSS configuration, and I would expect them to be mainly compile-time configurable from application point of view.

File: include/modem/location.h
Comment (jhirsi) [reply include/modem/location.h]: OK, good that ping is getting through and back, but that is only 28 bytes ICMP packet. I was concerned of actual longer cloud events and actual a-gps data, received MT data. I think that it is not guaranteed that the network is buffering that during window of priority mode and it might get lost, thus was thinking if usage of priority window could be limited to non-LTE phase.

File: lib/location/method_gnss.c
Comment (hiltunent) [reply lib/location/method_gnss.c:707]: changed

File: doc/nrf/releases/release-notes-changelog.rst
Comment (hiltunent) [reply doc/nrf/releases/release-notes-changelog.rst:510]: changed

File: applications/asset_tracker_v2/src/modules/location_module.c
Comment (hiltunent) [reply applications/asset_tracker_v2/src/modules/location_module.c]: I don't have a personal opinion, but if @trantanen says we will align configurability after NCS 2.2.0 I am tempted to leave it like this now.

File: include/modem/location.h
Comment (hiltunent) [reply include/modem/location.h:280]: It's a very nice text, but should it be here instead https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrfxlib/nrf_modem/doc/gnss_interface.html ?
We could then refer to that document. What do you think @peknis ?

Comment (trantanen) [reply include/modem/location.h:280]: Yep, it should be there actually.
I don't know how do we refer to there from the doxygen comments other than adding a link (which may get broken) to the text.

Comment (peknis) [reply include/modem/location.h:280]: I like the idea of moving longer texts to conceptual documentation instead of having it in API docs and then linking to it.

File: applications/asset_tracker_v2/src/modules/location_module.c
Comment (jhirsi) [reply applications/asset_tracker_v2/src/modules/location_module.c]: OK for me as well, although if keeping as runtime config, mosh location command support is needed as stated [in](https://github.com/nrfconnect/sdk-nrf/pull/9325#discussion_r1029099454).
In addition to mosh tests, I'd enable this after the PR merge in mosh_dev's ATv2 to testing feedback, because CIs are running there as well for ATv2. ping @moksanen.

File: include/modem/location.h
Comment (trantanen) [reply include/modem/location.h:280]: GNSS API mentions also that priority mode is max 40s.

Maybe we should just have send/receive data comments and 40s blocking time documented for the priority mode here. Simply because users are looking at location library abstraction and the less they need to follow the links the better. Some duplication of information is not always that bad.

File: include/modem/location.h
Comment (hiltunent) [reply include/modem/location.h]: That is a good point. It did not occur to me that the network might not be willing to buffer larger amounts of data. However, the timing is such in location lib that A-GPS and/or P-GPS data are usually downloaded first even before starting GNSS and therefore the priority mode should not interfere. @tokangas can confirm if this is still true as he was fiddling with the timing yesterday. 

That being said, I don't know if we can guarantee 100% that no data will be lost. On the other hand, the drawback of starting priority mode only when e.g. not registered to network would limit the usability of the feature.

File: include/modem/location.h
Comment (hiltunent) [reply include/modem/location.h:280]: Ok, I'll improve the comment here along the lines @trantanen proposed and let @peknis handle any other documents if need be.

Comment (hiltunent) [reply include/modem/location.h:287]: done and done

File: include/modem/location.h
Comment (trantanen) [inline include/modem/location.h:285]:
```suggestion
	 * If the device attempts to send data during the priority mode, it will be buffered and
	 * sent after the priority time window ends. In case of mobile terminated data reception
	 * during the priority mode, the network will typically buffer the data and send them to the
	 * device once the priority time window ends. However, it is possible that the network drops
	 * the data, or some protocol timer expires causing the data transfer to fail.
```

**Review by trantanen — APPROVED** [review]
OK for me now.

**Review by tokangas — APPROVED** [review]
Looks good to me.

**Review by jhirsi — APPROVED** [review]

**Review by peknis — APPROVED** [review]

**Pull request merged**
